grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java

```diff
@@ -321,22 +374,80 @@ public void run() {
         });
     }
 
+    @Override public void onError(final Throwable error) {
+      helper.runSerialized(new Runnable() {
+          @Override
+          public void run() {
+            handleStreamClosed(Status.fromThrowable(error)
+                .augmentDescription("Stream to GRPCLB LoadBalancer had an error"));
+          }
+        });
+    }
+
+    @Override public void onCompleted() {
+      helper.runSerialized(new Runnable() {
+          @Override
+          public void run() {
+            handleStreamClosed(
+                Status.UNAVAILABLE.withDescription("Stream to GRPCLB LoadBalancer was closed"));
+          }
+        });
+    }
+
+    // Following methods must be run in helper.runSerialized()
+
+    private void sendLoadReport() {
+      ClientStats stats = loadRecorder.generateLoadReport();
+      // TODO(zhangkun83): flow control?
+      try {
+        lbRequestWriter.onNext(LoadBalanceRequest.newBuilder().setClientStats(stats).build());
+        scheduleNextLoadReport();
+      } catch (Exception e) {
+        close(e);
+      }
+    }
+
+    private void scheduleNextLoadReport() {
+      stopLoadReporting();
+      if (loadReportIntervalMillis > 0) {
+        loadReportTask = timerService.schedule(
+            loadReportRunnable, loadReportIntervalMillis, TimeUnit.MILLISECONDS);
+      }
+    }
+
+    private void stopLoadReporting() {
+      if (loadReportTask != null) {
+        loadReportTask.cancel(false);
```

Comment:
There is a benign race here: if the timerService just started executing the task of `loadReportTask`,  namely line 346, but line 347 is not executed yet, it can not be canceled by `loadReportTask.cancel(false)`, then `sendLoadReport()` will be queued by `runSerialized`. In this case `cleanUp()` does not work.